src/gutter_tree.cpp

```diff
@@ -0,0 +1,598 @@
+      
+      parent_keys -= ceil(parent_keys/options);
+      options--;
+      buffers.push_back(bcb);
+      index++; // seperate variable because sometimes we skip stuff
+      if(bcb->is_leaf())
+        size += leaf_size + page_size; // leaves are of size == sketch
+      else 
+        size += buffer_size + page_size;
+    }
+  }
+
+    // allocate file space for all the nodes to prevent fragmentation
+  #ifdef LINUX_FALLOCATE
+    fallocate(backing_store, 0, 0, size); // linux only but fast
+  #endif
+    #ifdef WINDOWS_FILEAPI
+    // https://stackoverflow.com/questions/455297/creating-big-file-on-windows/455302#455302
+    // TODO: implement the above
+    throw std::runtime_error("Windows is not currently supported by GutterTree");
+  #endif
+  #ifdef POSIX_FCNTL
+    // taken from https://github.com/trbs/fallocate
+    fstore_t store_options = {
+      F_ALLOCATECONTIG,
+      F_PEOFPOSMODE,
+      0,
+      static_cast<off_t>(size),
+      0
+    };
+    fcntl(backing_store, F_PREALLOCATE, &store_options);
+  #endif
+    
+    backing_EOF = size;
+    // print_tree(buffers);
+}
+
+// serialize an update to a data location (should only be used for root I think)
+inline void GutterTree::serialize_update(char *dst, update_t src) {
+  node_id_t node1 = src.first;
+  node_id_t node2 = src.second;
+
+  memcpy(dst, &node1, sizeof(node_id_t));
+  memcpy(dst + sizeof(node_id_t), &node2, sizeof(node_id_t));
+}
+
+inline update_t GutterTree::deserialize_update(char *src) {
+  update_t dst;
+  dst.first  = *((node_id_t *) src);
+  dst.second = *((node_id_t *) (src + sizeof(node_id_t)));
+
+  return dst;
+}
+
+// copy two serailized updates between two locations
+inline void GutterTree::copy_serial(char *src, char *dst) {
+  memcpy(dst, src, serial_update_size);
+}
+
+/*
+ * Load a key from a given location
+ */
+inline node_id_t GutterTree::load_key(char *location) {
+  node_id_t key;
+  memcpy(&key, location, sizeof(node_id_t));
+  return key;
+}
+
+/*
+ * Helper function which determines which child we should flush to
+ */
+inline uint32_t which_child(node_id_t key, node_id_t min_key, node_id_t max_key, uint16_t options) {
+  node_id_t total = max_key - min_key + 1;
+  double div = (double)total / options;
+
+  uint32_t larger_kids = total % options;
+  uint32_t larger_count = larger_kids * ceil(div);
+  node_id_t idx = key - min_key;
+
+  if (idx >= larger_count)
+    return ((idx - larger_count) / (int)div) + larger_kids;
+  else
+    return idx / ceil(div);
+}
+
+/*
+ * Perform an insertion to the buffer-tree
+ * Insertions always go to the root
+ */
+insert_ret_t GutterTree::insert(const update_t &upd) {
+  // printf("inserting to gutter_tree . . . \n");
+  
+  // first calculate which of the roots we're inserting to
+  Node key = upd.first;
+  buffer_id_t r_id = which_child(key, 0, num_nodes-1, fanout); // TODO: Here we are assuming that num_nodes >= fanout
+  BufferControlBlock *root = buffers[r_id];
```

Comment:
for my understanding, there's no longer a single "root" buffer, rather we just insert directly to whichever of the (former) root's children the update is bound for?

Reply:
Correct